### PR TITLE
Add typed-protocols-doc

### PR DESCRIPTION
**Pull request by tdammers**
This adds a subproject that uses Template Haskell to generate HTML documentation from a typed protocol definition.

### CODE REVIEW

File: typed-protocols-doc/test/Network/TypedProtocol/Tests/ControlProtocol.hs [hunk 1]
```diff
@@ -0,0 +1,323 @@
+{-# LANGUAGE DataKinds #-}
```
Comment (coot) [inline typed-protocols-doc/test/Network/TypedProtocol/Tests/ControlProtocol.hs:1]: Is this the `kes` agent protocol, or just a variant for testing?

**Comment by coot** [comment]
We should build `typed-protocols-docs` in `GHA`, e.g. `.github/workflows/haskell.yml`.

File: cabal.project [hunk 2]
```diff
@@ -18,5 +18,12 @@ index-state:
 packages: ./typed-protocols
           ./typed-protocols-cborg
           ./typed-protocols-examples
+          ./typed-protocols-doc
+
+source-repository-package
+  type: git
+  location: https://github.com/input-output-hk/serdoc
+  tag: c1e6d6ae22d58a402ac359b26be7b922921d2a73
+  subdir: serdoc-core
```
[the remark this answers is not in the document]
Comment (tdammers) [reply cabal.project:27]: Would love to, yes.

File: typed-protocols-doc/test/Network/TypedProtocol/Tests/ControlProtocol.hs (hunk 1 above)
Comment (tdammers) [reply typed-protocols-doc/test/Network/TypedProtocol/Tests/ControlProtocol.hs:1]: It's an older version of one of the KES agent protocols, used here for testing. It was easier to just use this one than to come up with a made-up protocol that covers the same use cases.

**Review by coot — CHANGES_REQUESTED** [review]

File: typed-protocols-doc/typed-protocols-doc.cabal [hunk 3]
```diff
@@ -7,14 +7,22 @@ license:            Apache-2.0
 license-file:       LICENSE
 author:             Tobias Dammers
 maintainer:         tobias@well-typed.com
-copyright:          2023 IO Global
+copyright:          2023 Input Output Global Inc (IOG), 2023-2024 Intersect
```
[the remark this answers is not in the document]
Comment (tdammers) [reply typed-protocols-doc/typed-protocols-doc.cabal]: No problem.

**Comment by tdammers** [comment]
I've added typed-protocols-doc to the GHA workflow, and it's currently passing.

**Review by coot — APPROVED** [review]
LGTM

**Pull request merged**